src/server/multi_svr.c

```diff
@@ -295,14 +518,367 @@ gen_svr_inst_id(void)
+
+		/* broadcast resc usage */
+		if ((ret = send_resc_usage(mtfd, psvr_ru, 1, incr_ct)) != DIS_SUCCESS)
+			close_streams(mtfd, ret);
+
+		tpp_mcast_close(mtfd);
+		mtfd = -1;
+	}
+}
+
+/**
+ * @brief adding an alien node to the cache
+ * 
+ * @param[in,out] psvr - peer server where the node belongs to
+ * @param[in] pnode - alien node structure
+ */
+static void
+add_node_to_cache(server_t *psvr, pbs_node *pnode)
+{
+	if (!pnode)
+		return;
+
+	CLEAR_LINK(pnode->nd_link);
+	append_link(&psvr->mi_node_list, &pnode->nd_link, pnode);
+	pbs_idx_insert(alien_node_idx, pnode->nd_name, pnode);
+}
+
+/**
+ * @brief initialize the node from the batch status
+ * 
+ * @param[in] bstat - batch status 
+ * @param[in,out] psvr - peer server structure
+ * @return int
+ * @retval -1 : for error
+ */
+int
+init_node_from_bstat(struct batch_status *bstat, server_t *psvr)
+{
+	pbs_node *pnode;
+	pbs_list_head attrs;
+	struct batch_status *cur;
+
+	for (cur = bstat; cur; cur = cur->next) {
+		pnode = calloc(1, sizeof(pbs_node));
+		pnode->nd_name = strdup(cur->name);
+		pnode->nd_svrflags |= NODE_ALIEN;
+		copy_attrl_to_svrattrl(cur->attribs, &attrs);
+		if ((decode_attr_db(pnode, &attrs, node_attr_idx, node_attr_def, pnode->nd_attr, ND_ATR_LAST, 0)) != 0) {
+			log_errf(PBSE_INTERNAL, __func__, "Decode of node %s received from peer server has failed!",
+				 pnode->nd_name);
+			free(pnode);
+			free_attrlist(&attrs);
+			pbs_statfree(bstat);
+			return -1;
+		}
+
+		add_node_to_cache(psvr, pnode);
+		free_attrlist(&attrs);
+	}
+
+	pbs_statfree(bstat);
+	return 0;
+}
+
+/**
+ * @brief clear node from the cache and delete the node
+ * 
+ * @param[in,out] psvr - peer server struct
+ */
+static void
+clear_node_cache(server_t *psvr)
+{
+	pbs_node *pnode;
+	pbs_node *nd_next;
+
+	for (pnode = GET_NEXT(psvr->mi_node_list);
+	     pnode; pnode = nd_next) {
+		nd_next = GET_NEXT(pnode->nd_link);
+		CLEAR_LINK(pnode->nd_link);
+		pbs_idx_delete(alien_node_idx, pnode->nd_name);
+		free_pnode(pnode);
+	}
+}
+
+/**
+ * @brief clear the old node cache and update with new information
+ * 
+ * @param[in] stream - connection stream
+ * @param[in] bstat - batch status struct
+ * @return int 
+ * @retval -1 : for error
+ */
+static int
+update_node_cache(int stream, struct batch_status *bstat)
+{
+	server_t *psvr;
+
+	psvr = tfind2((u_long) stream, 0, &streams);
+	if (!psvr)
+		return -1;
+
+	log_eventf(PBSEVENT_DEBUG3, PBS_EVENTCLASS_SERVER, LOG_DEBUG, __func__,
+		   "node stat update received from server %s port %d",
+		   psvr->mi_host, psvr->mi_port);
+
+	/* DIFF_STAT TODO - this logic needs to be optimized after diff-stat */
+	clear_node_cache(psvr);
+
+	init_node_from_bstat(bstat, psvr);
+
+	return 0;
+}
+
+/**
+ * @brief
+ * 		find_alien_node() - find an alien node by its name
+ * @param[in]	nodename	- node being searched
+ *
+ * @return	strcut pbsnode *
+ * @retval	!NULL - success
+ * @retval	NULL  - failure
+ */
+pbs_node *
+find_alien_node(char *nodename)
```

Comment:
This function doesn't expect nodes in a special format. But I have kept the same checks which were present in find_nodebyname which was skipping over a leading parenthesis and terminate at a forward slash if they were any. Let me know if you think they will be no longer relevant.

Reply:
I think they are relevant. The Only thing is if we add an example it would be better to follow the code. In other words If it follows the same semantics as that of find_nobdebyname we can just mention the same in Doxygen header.